File: drive/src/main/java/de/mein/drive/sql/dao/FileDistTaskDao.java
package de.mein.drive.sql.dao;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

import de.mein.Lok;
import de.mein.auth.file.AFile;
import de.mein.auth.tools.N;
import de.mein.core.serialize.SerializableEntity;
import de.mein.core.serialize.deserialize.entity.SerializableEntityDeserializer;
import de.mein.core.serialize.serialize.fieldserializer.entity.SerializableEntitySerializer;
import de.mein.drive.bash.FsBashDetails;
import de.mein.drive.data.FileDistTaskWrapper;
import de.mein.drive.nio.FileDistributionTask;
import de.mein.sql.Dao;
import de.mein.sql.ISQLQueries;
import de.mein.sql.SqlQueriesException;

/**
 * You cannot put a JSON into a database on Android
 */
public class FileDistTaskDao extends Dao {
    private FileDistTaskWrapper dummy = new FileDistTaskWrapper();
    private FileDistTaskWrapper.FileWrapper dummyFW = new FileDistTaskWrapper.FileWrapper();

    public FileDistTaskDao(ISQLQueries sqlQueries) {
        super(sqlQueries, false);
    }


    public void insert(FileDistributionTask task) throws SqlQueriesException {
        N.r(() -> Lok.debug("INSERT COPY: " + SerializableEntitySerializer.serialize(task)));
        FileDistTaskWrapper wrapper = FileDistTaskWrapper.fromTask(task);
        Long id = sqlQueries.insert(wrapper);
        wrapper.getId().v(id);
        N.forEach(wrapper.getTargetWraps(), fileWrapper -> {
            sqlQueries.insert(fileWrapper);
        });
    }

    public boolean hasContent() throws SqlQueriesException {
        String query = "select count(1) from " + dummy.getTableName() + " where " + dummy.getDone().k() + "=?";
        Long count = sqlQueries.queryValue(query, Long.class, ISQLQueries.whereArgs(false));
        return count > 0;
    }

    public void markDone(long id) throws SqlQueriesException {
        String statement = "update " + dummy.getTableName() + " set " + dummy.getDone().k() + "=? where " + dummy.getId().k() + "=?";
        sqlQueries.execute(statement, ISQLQueries.whereArgs(true, id));
    }

    public Map<Long, FileDistributionTask> loadChunk() throws SqlQueriesException {
        String where = dummy.getDone().k() + "=? limit 10";
        Map<Long, FileDistributionTask> tasks = new HashMap<>();
        List<FileDistTaskWrapper> taskWrappers = sqlQueries.load(dummy.getAllAttributes(), dummy, where, ISQLQueries.whereArgs(false));
        N.forEach(taskWrappers, taskWrapper -> {
            FileDistributionTask task = new FileDistributionTask();
            task.setSourceHash(taskWrapper.getSourceHash().v());
            task.setServiceUuid(taskWrapper.getServiceUuid().v());
            task.setDeleteSource(taskWrapper.getDeleteSource().v());
            task.setSourceFile(AFile.instance(taskWrapper.getSourcePath().v()));
            if (taskWrapper.getSize().notNull()) {
                FsBashDetails bashDetails = (FsBashDetails) SerializableEntityDeserializer.deserialize(taskWrapper.getSourceDetails().v());
                task.setOptionals(bashDetails, taskWrapper.getSize().v());
            }

            String whereToo = dummyFW.getTaskId().k() + "=?";
            List<FileDistTaskWrapper.FileWrapper> fileWrappers = sqlQueries.load(dummyFW.getAllAttributes(), dummyFW, whereToo, ISQLQueries.whereArgs(taskWrapper.getId().v()));
            N.forEach(fileWrappers, fileWrapper -> {
                AFile targetFile = AFile.instance(fileWrapper.getTargetPath().v());
                task.addTargetFile(targetFile, fileWrapper.getTargetFsId().v());
            });
            task.initFromPaths();
            tasks.put(taskWrapper.getId().v(), task);
        });
        return tasks;
    }

    public void deleteMarkedDone() throws SqlQueriesException {
        String statement = "delete from " + dummy.getTableName() + " where " + dummy.getDone().k() + "=?";
        sqlQueries.execute(statement, ISQLQueries.whereArgs(true));
    }

    public void deleteAll() throws SqlQueriesException {
        sqlQueries.execute("delete from " + dummy.getTableName(), null);
    }

    public int countAll() throws SqlQueriesException {
        return sqlQueries.queryValue("select count(1) from " + dummy.getTableName(), Integer.class);

    }

    public int countDone() throws SqlQueriesException {
        return sqlQueries.queryValue("select count(1) from " + dummy.getTableName() + " where " + dummy.getDone().k() + "=?", Integer.class, ISQLQueries.whereArgs(true));
    }
}

File: drive/src/main/java/de/mein/drive/sql/dao/FileDistTaskDao.kt
package de.mein.drive.sql.dao

import java.util.HashMap

import de.mein.Lok
import de.mein.auth.file.AFile
import de.mein.auth.tools.N
import de.mein.core.serialize.SerializableEntity
import de.mein.core.serialize.deserialize.entity.SerializableEntityDeserializer
import de.mein.core.serialize.serialize.fieldserializer.entity.SerializableEntitySerializer
import de.mein.drive.bash.FsBashDetails
import de.mein.drive.data.FileDistTaskWrapper
import de.mein.drive.nio.FileDistributionTask
import de.mein.sql.Dao
import de.mein.sql.ISQLQueries
import de.mein.sql.SqlQueriesException

/**
 * You cannot put a JSON into a database on Android
 */
class FileDistTaskDao(sqlQueries: ISQLQueries) : Dao(sqlQueries, false) {


    private val dummy = FileDistTaskWrapper()
    private val dummyFW = FileDistTaskWrapper.FileWrapper()

    fun isComplete(): Boolean {
        val queryDifference = "select ((select count(1) from ${dummy.tableName})-(select count(1) from ${dummy.tableName} where ${dummy.state.k()}=?))"
        val count: Long = sqlQueries.queryValue(queryDifference, Long::class.java, ISQLQueries.whereArgs(FileDistributionTask.FileDistributionState.DONE))
        return count == 0L
    }


    @Throws(SqlQueriesException::class)
    fun insert(task: FileDistributionTask) {
        val wrapper = FileDistTaskWrapper.fromTask(task)
        val id = sqlQueries.insert(wrapper)
        wrapper.id.v(id)
        task.id = id
        N.forEach(wrapper.targetWraps) { fileWrapper -> sqlQueries.insert(fileWrapper) }
    }

    fun getNotReadyYetByHash(hash: String): FileDistributionTask? {
        val where = "${dummy.sourceHash.k()}=? and ${dummy.state.k()}=?"
        val wrapper = sqlQueries.loadFirstRow(dummy.allAttributes, dummy, where, ISQLQueries.whereArgs(hash, FileDistributionTask.FileDistributionState.NRY), FileDistTaskWrapper::class.java)

        return wrapperToTask(wrapper)
    }

    private fun wrapperToTask(taskWrapper: FileDistTaskWrapper?): FileDistributionTask? {
        if (taskWrapper == null)
            return null
        val task = FileDistributionTask()
        task.id = taskWrapper.id.v()
        task.sourceHash = taskWrapper.sourceHash.v()
        task.deleteSource = taskWrapper.deleteSource.v()
        if (taskWrapper.sourcePath.notNull())
            task.sourceFile = AFile.instance(taskWrapper.sourcePath.v())
        if (taskWrapper.size.notNull()) {
            val bashDetails = SerializableEntityDeserializer.deserialize(taskWrapper.sourceDetails.v()) as FsBashDetails
            task.setOptionals(bashDetails, taskWrapper.size.v())
        }

        val whereToo = dummyFW.taskId.k() + "=?"
        val fileWrappers = sqlQueries.load(dummyFW.allAttributes, dummyFW, whereToo, ISQLQueries.whereArgs(taskWrapper.id.v()))
        N.forEach(fileWrappers) { fileWrapper ->
            val targetFile = AFile.instance(fileWrapper.getTargetPath().v())
            task.addTargetFile(targetFile, fileWrapper.getTargetFsId().v())
        }
        task.initFromPaths()
        return task
    }

    fun completeJob(task: FileDistributionTask) {
        if (task.id == null) {
            insert(task)
            return
        }
        val wrapper = FileDistTaskWrapper.fromTask(task)
        wrapper.id.v(task.id)
        val where = "${dummy.id.k()}=?"
        N.forEach(wrapper.targetWraps) { fileWrapper ->
            fileWrapper.taskId.v(wrapper.id)
            sqlQueries.insert(fileWrapper)
        }
        sqlQueries.update(wrapper, where, ISQLQueries.whereArgs(task.id))
    }


    @Throws(SqlQueriesException::class)
    fun markDone(id: Long) {
        val statement = "update " + dummy.tableName + " set " + dummy.state.k() + "=? where " + dummy.id.k() + "=?"
        sqlQueries.execute(statement, ISQLQueries.whereArgs(FileDistributionTask.FileDistributionState.DONE, id))
    }

    @Throws(SqlQueriesException::class)
    fun loadChunk(): Map<Long, FileDistributionTask> {
        val where = dummy.state.k() + "=? limit 10"
        val tasks = HashMap<Long, FileDistributionTask>()
        val taskWrappers = sqlQueries.load(dummy.allAttributes, dummy, where, ISQLQueries.whereArgs(FileDistributionTask.FileDistributionState.READY))!!
        N.forEach(taskWrappers) { taskWrapper ->
            val task = wrapperToTask(taskWrapper)
            tasks[taskWrapper.id.v()] = task!!
        }
        return tasks
    }

    @Throws(SqlQueriesException::class)
    fun deleteMarkedDone() {
        val statement = "delete from " + dummy.tableName + " where " + dummy.state.k() + "=?"
        sqlQueries.execute(statement, ISQLQueries.whereArgs(FileDistributionTask.FileDistributionState.DONE))
    }

    @Throws(SqlQueriesException::class)
    fun deleteAll() {
        sqlQueries.execute("delete from " + dummy.tableName, null)
    }

    @Throws(SqlQueriesException::class)
    fun countAll(): Int {
        return sqlQueries.queryValue("select count(1) from " + dummy.tableName, Int::class.java)

    }

    @Throws(SqlQueriesException::class)
    fun countDone(): Int {
        return sqlQueries.queryValue("select count(1) from " + dummy.tableName + " where " + dummy.state.k() + "=?", Int::class.java, ISQLQueries.whereArgs(FileDistributionTask.FileDistributionState.DONE))
    }


    @Throws(SqlQueriesException::class)
    fun hasWork(): Boolean {
        val query = "select count(1) from " + dummy.tableName + " where " + dummy.state.k() + "=?"
        val count = sqlQueries.queryValue(query, Long::class.java, ISQLQueries.whereArgs(FileDistributionTask.FileDistributionState.READY))
        return count > 0
    }
}
